### PR TITLE
Reindex chunk

### PR DESCRIPTION
**Pull request by brong**
Quite tested now - it's been running very happily on production for a while :)  One bug which I'll push a commit for.  I don't think we need to rebase it all together.

### CODE REVIEW

**Review by rsto — COMMENTED** [review]
Agreed that the existing code is a bit unwieldy and this doesn't make it better. I wonder if we could hide the complexity one layer below in xapian_wrap. We add a DBW_BATCH flag to xapian_dbw_open and will it take of creating the interim databases. After the reindex we expect it all compacted to a single db for regular use anyway, right?

File: imap/jmap_mail_query.c [hunk 1]
```diff
@@ -957,7 +957,7 @@ HIDDEN matchmime_t *jmap_email_matchmime_init(const struct buf *mime, json_t **e
     const char *paths[2];
     paths[0] = matchmime->dbpath;
     paths[1] = NULL;
-    r = xapian_dbw_open(paths, &matchmime->dbw, 0);
+    r = xapian_dbw_open(paths, &matchmime->dbw, /*mode*/0, /*nosync*/1);
```
[the remark this answers is not in the document]
Comment (brong) [reply imap/jmap_mail_query.c:960]: Yeah, we could... I'll have a play with this one after I've got a test build out :)

[the remark this answers is not in the document]
Comment (brong) [reply imap/jmap_mail_query.c:960]: I think I'm happy to leave this for now.  We can add flags when we get to having three separate knobs :)

**Comment by brong** [comment]
@rsto I pushed one more commit plus rebased on master.  The top commit handles the case where there's nothing to write (we're repacking an empty DB for some reason - normally the user has deleted all their emails, sometimes just all their old emails that were indexed into the old file we're reindexing) - instead of bailing and aborting the repack we will just skip it.

File: imap/search_xapian.c [hunk 2]
```diff
@@ -2473,7 +2477,7 @@ static int end_message_update(search_text_receiver_t *rx, uint8_t indexlevel)
     if (!ptrarray_size(&tr->super.segs)) goto out;
 
     if (!tr->dbw) {
-        r = xapian_dbw_open((const char **)tr->activedirs->data, &tr->dbw, tr->mode);
+        r = xapian_dbw_open((const char **)tr->activedirs->data, &tr->dbw, tr->mode, /*nosync*/0);
```
Comment (rsto) [inline imap/search_xapian.c:2480]: we could add NOSYNC as an additional mode flag?

**Review by rsto — COMMENTED** [review]
Reading the source this looks good me. I haven't tested it, though, but AFAIK it's already running OK in a test environment.

**Pull request merged**